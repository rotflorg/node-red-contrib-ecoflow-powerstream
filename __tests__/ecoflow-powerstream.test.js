var helper = require('node-red-node-test-helper');
var testNode = require('../dist/ecoflow-powerstream.js');

helper.init(require.resolve('node-red'), {
  functionGlobalContext: { os:require('os') }
});

const TEST_MESSAGES = [
  [10,83,10,40,128,1,220,2,136,1,130,27,152,1,229,3,168,1,211,2,176,1,191,26,192,1,193,15,208,1,153,4,136,2,184,34,168,2,200,5,176,2,145,11,16,53,24,32,32,1,40,1,64,20,72,1,80,40,88,1,128,1,3,136,1,3,202,1,16,72,87,53,49,90,69,72,48,48,48,48,48,48,48,48,48,10,84,10,41,120,226,6,128,1,248,255,255,255,255,255,255,255,255,1,136,1,238,221,34,144,1,176,183,3,192,1,168,62,216,1,168,62,128,2,216,62,232,2,138,11,16,53,24,32,32,1,40,1,64,20,72,4,80,41,88,1,128,1,3,136,1,3,202,1,16,72,87,53,49,90,69,72,48,48,48,48,48,48,48,48,48],
  [10,75,10,32,128,1,215,2,136,1,233,26,168,1,215,2,176,1,239,26,232,1,228,248,255,255,255,255,255,255,255,1,176,2,138,11,16,53,24,32,32,1,40,1,64,20,72,1,80,32,88,1,128,1,3,136,1,3,202,1,16,72,87,53,49,90,69,72,48,48,48,48,48,48,48,48,48,10,63,10,20,120,203,6,136,1,152,223,34,192,1,165,62,128,2,210,62,232,2,129,11,16,53,24,32,32,1,40,1,64,20,72,4,80,20,88,1,128,1,3,136,1,3,202,1,16,72,87,53,49,90,69,72,48,48,48,48,48,48,48,48,48],
];

describe('ecoflow-powerstream Node', function () {

  beforeEach(function (done) {
    helper.startServer(done);
  })
  afterEach(function (done) {
    helper.unload();
    helper.stopServer(done);
  });

  it('should be loaded', function (done) {
    const flow = [{ id: 'n1', type: 'ecoflow-powerstream', name: 'test name' }];
    helper.load(testNode, flow, {}, function () {
      const n1 = helper.getNode('n1');
      expect(n1.name).toEqual('test name');
      done();
    });
  });

  it('should convert binary payload', function (done) {
    const flow = [{ id: 'n1', type: 'ecoflow-powerstream', name: 'test name', wires: [['n2']] }, { id: 'n2', type: 'helper' }];
    helper.load(testNode, flow, function () {
      const n2 = helper.getNode('n2');
      const n1 = helper.getNode('n1');
      const receivedMessages = [];
      n2.on('input', function (msg) {
        const copy = { ...msg };
        delete copy._msgid;
        receivedMessages.push(copy);
      });
      for (const binmsg of TEST_MESSAGES) {
        const buffer = Buffer.from(new Uint8Array(binmsg));
        n1.receive({ payload: buffer });
      }
      setTimeout(function() {
        expect(receivedMessages).toMatchSnapshot();
        done();
      }, 1000);
    });
  });
});
