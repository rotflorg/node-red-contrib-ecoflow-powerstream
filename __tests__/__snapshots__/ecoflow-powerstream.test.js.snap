// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecoflow-powerstream Node should convert binary payload 1`] = `
[
  {
    "payload": {
      "cmdFunc": 20,
      "cmdId": 1,
      "data": {
        "batInputVolt": 537,
        "invOutputCur": 712,
        "invOutputWatts": 1425,
        "llcOpVolt": 4408,
        "pv1InputVolt": 348,
        "pv1InputWatts": 485,
        "pv1OpVolt": 3458,
        "pv2InputVolt": 339,
        "pv2InputWatts": 1985,
        "pv2OpVolt": 3391,
      },
      "deviceSn": "HW51ZEH000000000",
      "msgType": "InverterHeartbeat",
      "payloadVer": 3,
      "version": 3,
    },
    "topic": "/ecoflow/powerstream",
  },
  {
    "payload": {
      "cmdFunc": 20,
      "cmdId": 4,
      "data": {
        "X_Unknown_15": 866,
        "X_Unknown_16": -8,
        "X_Unknown_17": 569070,
        "X_Unknown_18": 56240,
        "X_Unknown_24": 7976,
        "X_Unknown_27": 7976,
        "gridWatt_45": 1418,
        "uptime": 8024,
      },
      "deviceSn": "HW51ZEH000000000",
      "msgType": "InverterHeartbeat2",
      "payloadVer": 3,
      "version": 3,
    },
    "topic": "/ecoflow/powerstream",
  },
  {
    "payload": {
      "cmdFunc": 20,
      "cmdId": 1,
      "data": {
        "batInputWatts": -924,
        "invOutputWatts": 1418,
        "pv1InputVolt": 343,
        "pv1OpVolt": 3433,
        "pv2InputVolt": 343,
        "pv2OpVolt": 3439,
      },
      "deviceSn": "HW51ZEH000000000",
      "msgType": "InverterHeartbeat",
      "payloadVer": 3,
      "version": 3,
    },
    "topic": "/ecoflow/powerstream",
  },
  {
    "payload": {
      "cmdFunc": 20,
      "cmdId": 4,
      "data": {
        "X_Unknown_15": 843,
        "X_Unknown_17": 569240,
        "X_Unknown_24": 7973,
        "gridWatt_45": 1409,
        "uptime": 8018,
      },
      "deviceSn": "HW51ZEH000000000",
      "msgType": "InverterHeartbeat2",
      "payloadVer": 3,
      "version": 3,
    },
    "topic": "/ecoflow/powerstream",
  },
  {
    "payload": {
      "batInputVolt": 53.7,
      "batInputWatts": -92.4,
      "deviceSn": "HW51ZEH000000000",
      "invOutputWatts": 141.8,
      "pv1InputWatts": 48.5,
      "pv1OpVolt": 343.3,
      "pv2InputWatts": 198.5,
      "pv2OpVolt": 343.9,
      "pvInputWatts": 239.6,
    },
    "topic": "/ecoflow/powerstream/aggregated",
  },
]
`;
